File: shop.c
#include <stdio.h>
#include <string.h>
//struct is a way of combining pieces of data together
// C is not as advanced so it cannot refer to a struct unless it has processed it - Product Stock must come before customer as customer refers to it

struct Product {
    char* name;
    double price;
};

struct ProductStock{
    struct Product product; // must give it a name of a struct
    int quantity;
}; //ending the definition of a struct is a ; semi colon

struct Customer {
    char* name;
    double budget;
    struct ProductStock shoppingList[10];//the amount has to go in an array after the identifier which is the variable
    int index;
};

struct Shop{
    double cash;
    struct ProductStock stock[20];
};

void printProduct(struct Product p)
{
    printf("PRODUCT NAME: %s \nPRODUCT PRICE: %.2f\n", p.name, p.price);
    printf("----------\n");
}

void printCustomer(struct Customer c)
{
    printf("CUSTOMER NAME: %s \nCUSTOMER PRICE: %.2f\n", c.name,c.budget);
    printf("----------\n");
    for(int i = 0; i < c.index; i++)
    {
        printProduct(c.shoppingList[i].product);
        printf("%s ORDERS %d OF ABOVE PRODUCT\n", c.name, c.shoppingList[i].quantity);
        double cost = c.shoppingList[i].quantity * c.shoppingList[i].product.price; 
        printf("The cost to %s will be %.2f euro\n", c.name, cost);
    }
}
int main (void)
{
    struct Customer dominic = { "Dominic", 100.0};
    //printCustomer(dominic);

    struct Product coke = { "Can Coke", 1.10};
    struct Product bread = { "Bread", 0.70};
    // printProduct(coke);

    struct ProductStock cokeStock = { coke, 20 };
    struct ProductStock breadStock = { bread, 2 };

    dominic.shoppingList[dominic.index++] = cokeStock; // retrieve the vaule from index and increment it
    dominic.shoppingList[dominic.index++] = breadStock;

    printCustomer(dominic);
// printf ("The shop has %d 0f the product %s\n", cokeStock.quantity,cokeStock.product.name);

    return 0;
}
//when you right everything in the programme, the compilier knows how much it needs
//there is no memory management in this so far which is neccessary when you take in data from outside, you don't know the amount that will be taken in

File: G00376463-Martina-O'Brien-Assignment1/pythonshop-pp/shop.py
from dataclasses import dataclass
from typing import List
import csv

@dataclass
class Product:
    name: str
    price: float

@dataclass
class ProductStock:
    product: Product
    quantity: int

@dataclass
class Shop:
    cash: float
    stock: List[ProductStock]

@dataclass
class Customer:
    name: str
    budget: float
    shopping_list: List[ProductStock]

def create_and_stock_shop():
    s = Shop(200.0, [])
    with open('../stock.csv') as csv_file:
        csv_reader = csv.reader(csv_file, delimiter=',')
        line_count = 0
        for row in csv_reader:
            p = Product(row[0], float(row[1]))
            ps = ProductStock(p, float(row[2]))
            s.stock.append(ps)
            print(ps)
    return s

def print_product(p):
    print(f'\nPRODUCT NAME: {p.name} \nPRODUCT PRICE: {p.price} \n')

def print_customer(c):
    print(f'CUSTOMER NAME: {c.name} \nCUSTOMER BUDGET: {c.budget}')

    for item in c.shopping_list:
        print_product(item.product)

        print(f'{c.name} ORDERS {item.quantity} OF ABOVE PRODUCT')
        cost = item.quantity * item.product.price
        print(f'The cost to {c.name} will be €{cost}')

def print_shop(s):
    print(f'Shop has €{s.cash} in cash')
    for item in s.stock:
        print_product(item.product)
        print(f'The shop has {item.quantity} of the above')

p = Product("Coke", 1.0)
s = create_and_stock_shop()
print(s)